### PR TITLE
edit user should be editing the user specified, not the logged in user

### LINKED ISSUE #923: edit user should be editing the user specified, not the logged in user
from ```app/controllers/concerns/sufia/users_controller_behavior.rb```

```
def edit
    @user = current_user
    @trophies = @user.trophy_files
  end
```

should be

```
def edit
    @user = User.from_url_component(params[:id])
    @trophies = @user.trophy_files
  end
```

Permissions to edit other users is still enforced by CanCan

### PR DESCRIPTION
**Pull request by pgwillia**
fixes #923

### CODE REVIEW

**Comment by cam156** [comment]
@pgwillia You need to have some tests to be certain everyone can not edit other users.  I am not 100% certain cancan is checking the edit, or if there is an ability set up to default to only allowing you to edit your profile.

**Comment by pgwillia** [comment]
The tests are already there.  See [here](https://github.com/projecthydra/sufia/blob/master/spec/controllers/users_controller_spec.rb#L95)

**Comment by jcoyne** [comment]
@pgwillia there should probably be a check for `@user` (`assigns[:user]`) around here:
https://github.com/projecthydra/sufia/blob/master/spec/controllers/users_controller_spec.rb#L108-L109

The idea is to write a spec that fails without your change, but then after your change will pass.  That way if anyone reverts your code changes, they'll see a test fails.

**Comment by mjgiarlo** [comment]
This is ready for one last review with the new impl @pgwillia worked on. @jcoyne @awead etc.

**Comment by awead** [comment]
I'm :+1: but differ to @jcoyne has he originally commented on this.

File: spec/controllers/users_controller_spec.rb [hunk 1]
```diff
@@ -103,10 +103,24 @@
 
     context "when user attempts to edit another profile" do
       let(:another_user) { FactoryGirl.create(:user) }
-      it "redirects to show profile" do
-        get :edit, id: another_user.user_key
-        expect(response).to redirect_to(@routes.url_helpers.profile_path(another_user.to_param))
-        expect(flash[:alert]).to include("Permission denied: cannot access this page.")
+      context 'with default abilities' do
+        it "redirects to show profile" do
+          expect_any_instance_of(Ability).to receive(:can?).with(:edit, another_user).and_return(false)
```
Comment (mjgiarlo) [inline spec/controllers/users_controller_spec.rb]: I'm wondering if this test still passes without this line.

File: app/controllers/concerns/sufia/users_controller_behavior.rb [hunk 2]
```diff
@@ -142,4 +138,11 @@ def get_sort
            end
     return sort_val
   end
+
+  def deny_access(exception)
+    unless @user == current_user
+      redirect_to sufia.profile_path(@user.to_param), alert: "Permission denied: cannot access this page."
+  end
```
Comment (mjgiarlo) [inline app/controllers/concerns/sufia/users_controller_behavior.rb]: I think this line and the next each need two spaces prepended.

File: app/controllers/concerns/sufia/users_controller_behavior.rb [hunk 3]
```diff
@@ -142,4 +138,11 @@ def get_sort
            end
     return sort_val
   end
+
+  def deny_access(exception)
+    unless @user == current_user
```
Comment (mjgiarlo) [inline app/controllers/concerns/sufia/users_controller_behavior.rb]: What is this `unless` adding?  If we've already determined that the current user lacks the ability to edit/update `@user` on line 10, why the `unless`?

File: sufia-models/app/models/concerns/sufia/ability.rb [hunk 4]
```diff
@@ -23,7 +24,10 @@ def proxy_deposit_abilities
       can :reject, ProxyDepositRequest, receiving_user_id: current_user.id, status: 'pending'
       # a user who sent a proxy deposit request can cancel it if it's pending.
       can :destroy, ProxyDepositRequest, sending_user_id: current_user.id, status: 'pending'
-      can :edit, ::User, id: current_user.id
+    end
+
+    def user_abilities
+      can [:edit, :update, :toggle_trophy], ::User, id: current_user.id
```
Comment (mjgiarlo) [inline sufia-models/app/models/concerns/sufia/ability.rb]: Looks good, and this is what you'll be overriding in HydraNorth, right?

File: app/controllers/concerns/sufia/users_controller_behavior.rb (hunk 3 above)
Comment (pgwillia) [inline app/controllers/concerns/sufia/users_controller_behavior.rb]: You're right, I'll change it.  Got ahead of myself by thinking about doing the same thing to user_not_current_user in which case deny_access would need a conditional for the two user cases.

File: app/controllers/concerns/sufia/users_controller_behavior.rb (hunk 2 above)
Comment (pgwillia) [inline app/controllers/concerns/sufia/users_controller_behavior.rb]: +1

File: spec/controllers/users_controller_spec.rb (hunk 1 above)
Comment (pgwillia) [inline spec/controllers/users_controller_spec.rb]: Yes this test still passes without this line. expect_any_instance_of is an expect statement.  It is not mocking Ability. It is asserting the default abilities.

File: sufia-models/app/models/concerns/sufia/ability.rb (hunk 4 above)
Comment (pgwillia) [inline sufia-models/app/models/concerns/sufia/ability.rb]: Right!

File: app/controllers/concerns/sufia/users_controller_behavior.rb (hunk 3 above)
Comment (mjgiarlo) [inline app/controllers/concerns/sufia/users_controller_behavior.rb]: OK!

File: spec/controllers/users_controller_spec.rb (hunk 1 above)
Comment (mjgiarlo) [inline spec/controllers/users_controller_spec.rb]: OK, very good!

File: sufia-models/app/models/concerns/sufia/ability.rb (hunk 4 above)
Comment (mjgiarlo) [inline sufia-models/app/models/concerns/sufia/ability.rb]: Awesome. When you force-push your changes, feel free to ping in #projecthydra.

**Comment by mjgiarlo** [comment]
Restarting the build.

**Comment by mjgiarlo** [comment]
@pgwillia Could you also squash your commits and force-push again in the meantime?

**Comment by mjgiarlo** [comment]
Note that the Travis failures above are unrelated to this PR, so this can be merged once @pgwillia squashes and force-pushes again.

**Comment by mjgiarlo** [comment]
:+1: :clap: 

Thanks, @pgwillia!

**Comment by awead** [comment]
Huzzah! :maple_leaf: :maple_leaf: :maple_leaf:

**Pull request merged**